java-checks/src/main/java/org/sonar/java/checks/SpecializedFunctionalInterfacesCheck.java

```diff
@@ -0,0 +1,239 @@
+      matchFunctionalInterface(interfaceType)
+        .ifPresent(treeAndStringInterface -> reportIssue(treeAndStringInterface.classInterface, REPORT_STRING + " '" + treeAndStringInterface.reportString + "'"));
+    }
+  }
+
+  private static boolean isAnonymousClass(ExpressionTree initializeTree) {
+    return initializeTree.is(Tree.Kind.NEW_CLASS) && ((NewClassTree) initializeTree).classBody() != null;
+  }
+
+  private static Optional<InterfaceTreeAndStringPairReport> matchFunctionalInterface(TypeTree interfaceType) {
+    Optional<TypeArguments> args = interfaceParameters(interfaceType);
+    if (args.isPresent()) {
+      TypeArguments arguments = args.get();
+      Optional<String> reportString;
+      switch (interfaceType.symbolType().fullyQualifiedName()) {
+        case "java.util.function.Function":
+          reportString = handleFunctionInterface(arguments);
+          break;
+        case "java.util.function.BiFunction":
+          reportString = handleBiFunctionInterface(arguments);
+          break;
+        case "java.util.function.BiConsumer":
+          reportString = handleBiConsumerInterface(arguments);
+          break;
+        case "java.util.function.Supplier":
+        case "java.util.function.Consumer":
+        case "java.util.function.Predicate":
+        case "java.util.function.UnaryOperator":
+        case "java.util.function.BinaryOperator":
+          reportString = handleOneParameterInterface(arguments).map(s -> s + interfaceType.symbolType().name());
+          break;
+        default:
+          return Optional.empty();
+      }
+      if (reportString.isPresent()) {
+        return Optional.of(new InterfaceTreeAndStringPairReport(reportString.get(), interfaceType));
+      }
+    }
+    return Optional.empty();
+  }
+
+  private static Optional<String> handleOneParameterInterface(TypeArguments args) {
+    String reportString = new ParameterTypeNameAndTreeType(args.get(0)).paramTypeName;
+    if (reportString != null) {
+      return Optional.of(reportString);
+    }
+    return Optional.empty();
+  }
+
+  private static Optional<String> handleFunctionInterface(TypeArguments args) {
+    ParameterTypeNameAndTreeType firstArgument = new ParameterTypeNameAndTreeType(args.get(0));
+    ParameterTypeNameAndTreeType secondArgument = new ParameterTypeNameAndTreeType(args.get(1));
+    String reportString = null;
+    if (firstArgument.paramType.equals(secondArgument.paramType)) {
+      reportString = "UnaryOperator<" + firstArgument.paramType.name() + ">";
+    } else if (firstArgument.paramTypeName != null && secondArgument.paramTypeName != null) {
+      reportString = firstArgument.paramTypeName + "To" + secondArgument.paramTypeName + "Function";
+    } else if (firstArgument.paramTypeName == null && secondArgument.paramTypeName != null) {
+      reportString = "To" + secondArgument.paramTypeName + "Function<" + firstArgument.paramType.name() + ">";
```

Comment:
`String.format("To%sFunction<%s>", secondArgument.paramTypeName, firstArgument.paramTypeName)`